python/cuml/cuml/manifold/t_sne.pyx

```diff
@@ -141,6 +132,128 @@ _SUPPORTED_METRICS = {
 }
 
+_SUPPORTED_METHODS = {
+    "barnes_hut": TSNE_ALGORITHM.BARNES_HUT,
+    "exact": TSNE_ALGORITHM.EXACT,
+    "fft": TSNE_ALGORITHM.FFT,
+}
+
+_SUPPORTED_INITS = {
+    "random": TSNE_INIT.RANDOM,
+    "pca": TSNE_INIT.PCA,
+}
+
+
+def _check_numeric(estimator, name, gt=None, ge=None, lt=None, le=None):
+    """Check that a numeric parameter `name` is within valid bounds"""
+    value = getattr(estimator, name)
+    cls_name = type(estimator).__name__
+    if gt is not None and value <= gt:
+        raise ValueError(f"{cls_name} requires `{name} > {gt}`, got {value}")
+    if ge is not None and value < ge:
+        raise ValueError(f"{cls_name} requires `{name} >= {ge}`, got {value}")
+    if lt is not None and value >= lt:
+        raise ValueError(f"{cls_name} requires `{name} < {lt}`, got {value}")
+    if le is not None and value > le:
+        raise ValueError(f"{cls_name} requires `{name} <= {le}`, got {value}")
+    return value
+
+
+def _check_mapping(estimator, name, mapping):
+    """Check that a parameter `name` contained within a valid mapping"""
+    value = getattr(estimator, name)
+    cls_name = type(estimator).__name__
+    try:
+        return mapping[value]
+    except KeyError:
+        raise ValueError(
+            f"{cls_name} expects `{name}` to be one of {sorted(mapping)}, got {value}"
+        ) from None
+
+
+cdef _init_params(self, int n_samples, TSNEParams &params):
+    """Validate TSNE parameters and initialize a TSNEParams instance."""
+    if (n_components := self.n_components) != 2:
+        raise ValueError(
+            f"Currently TSNE only supports n_components = 2, got {self.n_components}"
+        )
+    perplexity = _check_numeric(self, "perplexity", gt=0)
+    early_exaggeration = _check_numeric(self, "early_exaggeration", ge=0)
+    late_exaggeration = _check_numeric(self, "late_exaggeration", ge=0)
+    learning_rate = _check_numeric(self, "learning_rate", gt=0)
+    adaptive_learning = _check_mapping(
+        self, "learning_rate_method", {"adaptive": True, "none": False, None: False}
+    )
+    n_iter = _check_numeric(self, "n_iter", gt=0)
+    min_grad_norm = _check_numeric(self, "min_grad_norm", ge=0)
+    angle = _check_numeric(self, "angle", ge=0, le=1)
+    n_neighbors = _check_numeric(self, "n_neighbors", gt=0)
+    perplexity_max_iter = _check_numeric(self, "perplexity_max_iter", ge=0)
+    exaggeration_iter = _check_numeric(self, "exaggeration_iter", ge=0)
+    pre_momentum = _check_numeric(self, "pre_momentum", gt=0, lt=1)
+    post_momentum = _check_numeric(self, "post_momentum", gt=0, lt=1)
+    init = _check_mapping(self, "init", _SUPPORTED_INITS)
+    algo = _check_mapping(self, "method", _SUPPORTED_METHODS)
+    metric = _check_mapping(self, "metric", _SUPPORTED_METRICS)
+
+    if n_samples < 2:
+        raise ValueError("TSNE requires >= 2 samples")
+
+    exaggeration_iter = min(exaggeration_iter, self.n_iter)
+    n_neighbors = min(n_neighbors, 1023, n_samples)
```

Comment:
Looks like we should be issuing a warning here if we want to maintain the previous behavior.

Reply:
I decided this warning wasn't actionable by the user and wasn't worth it. Happy to add it back in if you think it's necessary.

Reply:
It's actionable in the sense that the user can set it to <1024. I think the "correct" behavior would actually be to fail with a `ValueErrorI`. I suggest that we maintain the current behavior (issuing a warning) and revisit this in an potential future API overhaul, i.e., let's keep this PR non-breaking.

Reply:
Done